src/Sulu/Bundle/ContentBundle/Search/EventListener/ReindexListener.php

```diff
@@ -94,59 +111,121 @@ public function onIndexRebuild(IndexRebuildEvent $event)
             'SELECT * FROM [nt:unstructured] AS a WHERE ' . $condition
         );
 
-        $count = [];
+        $batchSize = 50;
+        $offset = $this->resumeManager->getCheckpoint(self::CHECKPOINT_NAME, 0);
+
+        // TODO: make this an option.
+        $batchLimit = null;
+        $batchCount = 0;
+
+        // index in batches: indexing in batches means:
+        //
+        //   1. Jackalope does not need to hydrate all of the PHPCR nodes from the query.
+        //   2. We can clear the document manager registry periodically.
+        //
+        // neither of the above points drastically reduce memory usage.
+        do {
+            $query->setMaxResults($batchSize);
+            $query->setFirstResult($offset);
+            $documents = $query->execute();
+
+            if ($documents->count() === 0) {
+                break;
+            }
+
+            $output->write(PHP_EOL);
+            $output->writeln(' <info>offset</info> ' . $offset);
+            $result = $this->indexBatch($output, $filter, $documents);
+            $output->write(PHP_EOL);
+
+            // this reduces overall memory usage by 1% in one experiment, so is not essential
+            $this->documentManager->clear();
+            $offset += $batchSize;
+            ++$batchCount;
+
+            if (null !== $batchLimit && $batchCount > $batchLimit) {
+                break;
+            }
+            $this->resumeManager->setCheckpoint(self::CHECKPOINT_NAME, $offset);
```

Comment:
@webmozart The crucial part for me was that this class had to use the `ResumeManager` on its own. So if the user who extends the MassiveSearchBundle this way forgets about that, the resuming doesn't work as expected.